fork-management-for-infra-providers/README.md

```diff
@@ -0,0 +1,148 @@
+# Fork Management for CAPI+Infra controllers
+
+The purpose of this RFC is to specify how we will manage forked repositories for CAPI and infra providers such as CAPVCD, CAPA, CAPO etc.
+
+## Problem Statement
+
+We use open-source Kubernetes operators in our platform and we need to customize them at some points.
+
+- We deploy them via the app platform so we need to bundle them as apps.
+- We have our requirements regarding deployment/configuration specs (e.g. common labels, CRD management etc.) so we need to customize the manifests.
+- We sometimes need to maintain some diffs in the core code since
+  - We need hotfixes and we don't have time to wait for merges in upstream repositories.
+  - We have a pretty specific temporary/permanent requirement and it is not possible to add it to upstream repositories. We must fight for keeping this diff as small/short-term as possible. 
+
+We need to 
+- define a fork management policy including repo management and branching strategies
+- automate code syncs, image building and releases 
+
+## Current status
+
+### Repositories
+
+We have 2 repositories per operator
+1. Fork of the upstream repository 
+- Examples:
+  - https://github.com/giantswarm/cluster-api
+  - https://github.com/giantswarm/cluster-api-provider-aws
+  - https://github.com/giantswarm/cluster-api-provider-cloud-director
+  - https://github.com/giantswarm/cluster-api-provider-gcp
+
+1. A repository for bundling the operator as an app 
+- Naming: https://github.com/giantswarm/\<fork-name\>-app
+- Examples
+  - https://github.com/giantswarm/cluster-api-app
+  - https://github.com/giantswarm/cluster-api-provider-aws-app
+  - https://github.com/giantswarm/cluster-api-provider-cloud-director-app
+  - https://github.com/giantswarm/cluster-api-provider-gcp-app
+
+### Syncing fork repositories
+
+There is no common way.
+
+In CAPVCD, since we have some diffs between our fork and upstream repo, we have a mechanism as follows:
+- `main` branch of the repo is the same as the upstream one. We sync this branch by clicking `Sync fork` button in the UI.
+- `giantswarm-main` branch of the repo contains our changes. We manually rebase it on top of `main` branch when necessary and open a new PR each time. The developer is supposed to resolve conflicts during rebasing.
+ 
+### Bundling + Customization
+
+There are some differences between repositories but the general structure in app repositories is as follows:
+
+- We have Makefile + Bash scripts to
+  - Check out a specific commit/branch of fork/upstream repo. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L14)
```

Comment:
We have a similar problem when updating any external chart from upstream. We're adapting git to solve all our needs, including git's subtree where needed. That way we don't need any synchronization scripts, as every change we do is saved in git itself.

We have docs about how it works here: https://handbook.giantswarm.io/product/managed-apps/dev-experience/git-subtree/

I also did a few demos that show how it works in real life:
- https://drive.google.com/file/d/1T9j2-sk3CEdp1-7_Q00VHj3s11eK6_jG/view?usp=share_link
- https://drive.google.com/file/d/11BWAOIm6Mr04DRPf-E4RHf_y7QSee5f8/view?usp=share_link
- https://drive.google.com/file/d/1xb8MGhv5OiBgwL5KMCZCRG3cw59xYOtT/view?usp=share_link

Reply:
@piontec do we build images in fork repositories in this structure or do we consume the upstream images directly? I didn't see anything about image building.

Reply:
> We have a similar problem when updating any external chart from upstream. We're adapting git to solve all our needs, including git's subtree where needed. That way we don't need any synchronization scripts, as every change we do is saved in git itself.
> 
> We have docs about how it works here: https://handbook.giantswarm.io/product/managed-apps/dev-experience/git-subtree/
> 
> I also did a few demos that show how it works in real life:
> 
> * https://drive.google.com/file/d/1T9j2-sk3CEdp1-7_Q00VHj3s11eK6_jG/view?usp=share_link
> * https://drive.google.com/file/d/11BWAOIm6Mr04DRPf-E4RHf_y7QSee5f8/view?usp=share_link
> * https://drive.google.com/file/d/1xb8MGhv5OiBgwL5KMCZCRG3cw59xYOtT/view?usp=share_link

isn't that a different scenario? There are no upstream Helm charts for CAPx.